### ISSUE TITLE
Cannot Mint Custom JWT in Python

### ISSUE THREAD

**Issue by swamidass** [supabase/supabase#26716]
# Bug report

- [ x] I confirm this is a bug with Supabase, not with my own application. 
- [ x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Supabase does not accept any modifications to the tokens 

## To Reproduce

See these snippets, using your own configuration:

```python
import jwt # pyjwt library
from supabase import create client

key =  # insert service key from supabase UI
secrete = # insert jwt secret from supabase UI
url = # insert url
```

This key works just fine with the client.

```python
# works!
supabase.create_client(my_key).table("mytable").select("*").limit(5).execute()
```

And we can inspect the claims using pyjwt:

```python
claims = jwt.decode(key,secret, algorithms=["HS256"]) 
```
```json
{'iss': 'supabase',
'ref':  '<redacted>',
'role': 'service_role',
'iat': 1702682963,
'exp': 2018258963}
````
Pyjwt can take these exact same claims to create an identical key, which validates that it is working just fine.

```python
my_key =  jwt.encode(claims, secret, algorithm="HS256")
assert key == my_key

# works just fine
supabase.create_client(my_key).table("mytable").select("*").limit(5).execute()
```

But even the most minor change to the claims invalidates this process. For example, if we round off the exp a bit (which should not impact auth at all), the key no longer works.

```python
claims = {'iss': 'supabase',
 'ref': '<redacted>',
 'role': 'service_role',
 'iat': 1702682963,
 'exp': 2018258900} # note the change to the 'exp'
}

my_key =  jwt.encode(claims, secret, algorithm="HS256")

# Does NOT work. 
supabase.create_client(my_key).table("mytable").select("*").limit(5).execute()
# APIError: {'message': 'Invalid API key', 'hint': 'Double check your Supabase `anon` or `service_role` API key.'}
```
Other small changes (I tried several), also break the key. In this case, I just added a 'test' claim. 

```
# now a small change
claims = {'iss': 'supabase',
 'ref': '<redacted>',
 'role': 'service_role',
 'iat': 1702682963,
 'exp': 2018258963},
'test': 1. # new key
}

my_key =  jwt.encode(claims, secret, algorithm="HS256")

# Does NOT work. 
supabase.create_client(my_key).table("mytable").select("*").limit(5).execute()
# APIError: {'message': 'Invalid API key', 'hint': 'Double check your Supabase `anon` or `service_role` API key.'}
```

This problem arises with really _any_ change made to the key's claims. So it is not possible to create custom JWTs.

## Expected behavior

Expected behavior, consistent with documentation, custom claims can be added to keys and still work on the API.

## System information

- OS:  macOS
- Version of supabase-js: [e.g. 2.4.5]

**Comment by swamidass** [comment]
Note that the same problem arises with the anon key.

**Comment by swamidass** [comment]
Thanks for help. I do not really understand the difference between "API Key" and "JWT Token used for Authorization". As I understand it, the supabase client uses an api key that is a jwt token to authorize?  

But I suppose you are saying that at some point, when using a key with create_client, the system does an exact character for character lookup of the key somehow. Is that correct? If so, that isn't documented any where I could find (with documentation that seems to say something quite different), and might explain the issue. But it still leaves me with some questions on how to solve my use case.

## What We Are Trying to Do

Right now our user authentication is handled by cognito, not supabase, and we can authenticate users through the OATH2 flow at our website, without touching supabase. Once we authenticate them, I want to grant each user a key to use that encodes a few details (e.g. their email), which we can then use for RLS as they make requests to supabase. 

No users are currently included in the supabase auth users table, and it seems like it would be really challenging to change this. Maybe their could be a migration path to this down the line, but I'm not sure. We could include a single "webui" user (or something like that), as long as we can dynamically add different claims each time. In that case, I'm not sure a hook will work.

## The Key Questions

So what approach should we take to this?  Perhaps these are the key questions:

Outside of supabase, how can I dynamically add different claims to the same supabase user in a way that supabase can see these claims and use them for RLS?

Or is their a better way to architect this?

**Comment by j4w8n** [comment]
Thanks for the explanation @swamidass. You can definitely use custom JWTs for this.

I think the challenge would be deciding when to set the JWT to expire. If you set it too short, it might stop working while the user is still logged-in. If too long, then if the JWT is compromised, it could be used to access that user's data. And in this scenario, there's no way to refresh the JWT with a refresh token. I suppose you could create a just-in-time JWT for each Supabase request that a logged-in user had; and set the expiration for like two minutes in the future. You'd need to do this on the server-side, since you'd be using your project's JWT secret to sign the JWTs.

Then, you could setup an RLS policy that checks whatever unique claims you put in the JWT (e.g. email).

**Comment by swamidass** [comment]
I can manage the refresh through the UI backend.  That isn't hard in the use case I'm envisioning.

I'm still unclear on how to actually create a key that will work with API, and how to actually attached it to the supabase client. Can you point me to any example snippets or documentation?

**Comment by j4w8n** [comment]
I'm not familiar with the Python implementation of Supabase. It [looks like](https://github.com/supabase-community/supabase-py/blob/main/supabase/lib/client_options.py) `create_client` has a `headers` option though, so you'd set the `Authorization` header to `Bearer <jwt-value>`. This is enough to make Supabase calls as the user. However, things like `get_session()`, `get_user()`, and `refresh_session()` will not work because the Supabase client is not aware of any session - that is, if the Python implementation works similarly to the Javascript one.

[This](https://supabase.com/docs/reference/python/initializing) initializing page has a "With timeout option" tab, in case you're not familiar with how to attach the `headers` option to the client.

**Comment by swamidass** [comment]
Okay, it seems that I now have some working code. In python (or nodejs), first make the key.

```
my_key =  jwt.encode(claims, secret, algorithm="HS256")
```
Then make a supabase client that has this key in the auth header, but uses either the anon or service_role key,

```
supabase = create_client(
    url,
    service_key,
    ClientOptions(headers={"Authorization": f"Bearer {my_key}"}),
)
```
Now postgres see the claims just fine. These claims can be really anything you like. Moreover, 

1. Supabase will respect the "role" claim, which you can set to authenticated, service_role, anon, or anything else.
2. Supabase will respect the "exp" claim, which you can either leave out for a non-expiring key.

This works painlessly. I just wish it was documented somewhere!!!

**Comment by j4w8n** [comment]
Awesome!

I don't believe you'd want to use the service role key in these situations, as it bypasses RLS - only use the anon key.

**Comment by swamidass** [comment]
I checked. Turns out that even if you use the service_role API-key, it will not bypass RLS, and instead use just the header key for RLS with role="authenticated" or "anon", ignoring service key.  

Remarkably, you can also use the anon-key as the API-key, but if the header-key has role="service_role", then even with the anon key, you will bypass RLS. You don't actually need the service_role API-key to bypass RLS. 

Of course, you wouldn't want to reveal your service key publicly. Instead, for example, in the browser you'd pass along the anon-key. The JWT secret is the key that unlocks everything, it seems.

**Comment by j4w8n** [comment]
If this is the case, then I'd swap "perfectly fine" with "unneeded" and use the anon key. I'd say best practice is to only use the service role key when needed.
> I checked. Turns out that even if you use the service_role API-key, it will NOT bypass RLS, and instead use just the header key for RLS with role="authenticated" or "anon", ignoring service key. So that means it's perfectly fine to use the service_role API-key server-side in this way.

Yeah, this is expected; as the `role` is important on the backend. As you've discovered, you can create your own service role JWTs.
> Remarkably, you can also use the anon-key as the API-key, but if the header-key has role="service_role", then even with the anon key, you will bypass RLS. You don't actually need the service_role API-key to bypass RLS.

Agreed.
> Of course, you wouldn't want to reveal your service key publicly. Instead, for example, in the browser you'd pass along the anon-key. The JWT secret is the key that unlocks everything, it seems.

Looks like you've got a solution! One thing to keep your eye on is [this PR](https://github.com/supabase/supabase-js/pull/1004#pullrequestreview-2079597246). I forgot about it, until someone randomly approved it this morning. I'm not sure if/when it'll drop.

**Comment by swamidass** [comment]
So it seems that I have an additional problem, this time in nodejs. I need to be able to use multiple different access keys in a given process. That creates a problem, becase:

```javascript
  const client1 = createClient<Database>(url, public_key, {
    global: { headers: {Authorization: `Bearer ${key1}`}}
    auth: {
      autoRefreshToken: false,
      persistSession: true,
      detectSessionInUrl: true,
    },
  });
  const client2 = createClient<Database>(url, public_key, {
    global: { headers: {Authorization: `Bearer ${key2}`}}
    auth: {
      autoRefreshToken: false,
      persistSession: true,
      detectSessionInUrl: true,
    },
  });
```

Produces this warning:

```
Multiple GoTrueClient instances detected in the same browser context. It is not an error, but this should be avoided as it may produce undefined behavior when used concurrently under the same storage key.
```

I don't get this error in python. 

What is the correct way to specify the access key? Is there a flag I can pass to ensure that gotrue isn't getting initialized? Or a way to alter the headers on a singleton supabase client so that each call can have a different access key?


So how are we suppoed

**Comment by j4w8n** [comment]
I'm not sure why you're getting that warning in node - unless you're using something like jsdom, that creates `document`.

Nonetheless, you should be fine with what you're doing as long as you're not logging in any users with these clients - implying you don't have to worry about the clients fighting over the same storage object, as the log warns about.

As an aside: if this is truly being done on the server side, you can change `detectSessionInUrl` to `false` as well.

**Comment by swamidass** [comment]
Oh, dang. I should have set that to false. I'll let you know if that didn't work.

**Comment by swamidass** [comment]
Alright, so with this code:

```javascript
  createClient<Database>(url, public_key, {
    auth: {
      autoRefreshToken: false,
      persistSession: false,
      detectSessionInUrl: false,
    },
  });
  createClient<Database>(url, public_key, {
    auth: {
      autoRefreshToken: false,
      persistSession: false,
      detectSessionInUrl: false,
    },
  });

```
I still get the warning. And to clarify, this is in the browser.

**Comment by j4w8n** [comment]
Gotcha. If you're in the browser, then yeah, that warning will throw if you create multiple browser clients. Even with that warning, what you're doing shouldn't be an issue, since you're not storing any sessions in storage (via traditional logins or a call to `setSession()`) - or at least I assume you're not, since you're manually setting the authorization header.

**Comment by Hallidayo** [comment]
Hi @swamidass - Just checking in here, did you get this all working?

**Comment by jzhou891** [comment]
@Hallidayo Hey just an FYI I tried doing this in python and the minted token fails to pass RLS with the anon key but works with the service key. I looked into the code, and the reason why is because the session doesn't get initialized, which means the `_get_token_header(self)` function in the `client.py` file for `SyncClient` falls back to the `supabase_key`. (in this case the anon key) that is provided during client initializing. From skimming the code it looked like the session never got set based on the headers that got passed in.

I was able to get it to work by simply authing the created client with the minted token using `postgrest.auth` but I still feel like this logic should be baked into the constructor. Would be open to making a PR in that case.
